api/api.go

```diff
@@ -587,8 +594,15 @@ func (r *request) toHTTP() (*http.Request, error) {
 		}
 	}
 
+	ctx := func() context.Context {
+		if r.ctx != nil {
+			return r.ctx
+		}
+		return context.Background()
```

Comment:
Is this reachable? The `request` should have `setQueryOptions` set already. If it can be reached, do we need a closure here, instead of just assigning the variable conditionally?

Reply:
I wasn't sure if it was a guarantee that `setQueryOptions` was called and wanted to avoid a null reference.  
For example, places that call `setWriteOptions` do not also call `setQueryOptions`. I see 2 ways to proceed
1. Also add it to setWriteOptions and then assume it will be always set, risking a null exception if anyone calls `newRequest` without `setWriteOptions`
2. leave the conditional. FWIW consul treats it as a condition https://github.com/hashicorp/consul/blob/master/api/api.go#L834 - but I do not like that it calls WithContext only sometimes, I prefer to run the same code all the time where I can :) 

If we leave it we can remove the IIF if you prefer for style. I used it here because it becomes clear exactly how `ctx` is set and you don't have to worry about future code edits spattering around flow control that changes ctx on different conditions - all paths will go in one place. 

Just let me know what you want to do

Reply:
Oh right, I was forgetting that `toHTTP()` was being used by writes as well.

Here's what I'd propose then, which will resolve my other questions about the API as well:
* Let's add a `ctx` field and `WithContext()`/`Context()` methods to `api.WriteOptions`, just as Consul has done. This makes the right thing the obvious thing for API consumers.
* Add a `r.ctx = q.Context()` line to `setWriteOptions`  
* And then sure, let's just keep extra closure as-is. Maybe it'll get pulled out to a `context()` method on `request` in some later refactor if we think that's useful down the road.

cc'ing in @cgbaker: does that API sound objectionable to you?

Reply:
yeah, i think so. we don't have a particular need for cancelling writes, mainly cancelling blocking queries.

Reply:
Updated the PR. It's still probably good on write queries - imagine the destination is unavailable and the app is waiting on TCP timeout and wants to cancel